### PR TITLE
Fixes text shadows for v5, take 2

### PR DESCRIPTION
**Pull request by themoonrat**
Alternative to https://github.com/pixijs/pixi.js/pull/5943

Goes back to v4 style of separate draw calls to make sure all of the drop shadow is underneath all of the text; but this time rather than using the first drawn text as the actual drop shadows, it draws this text off screen, whilst positioning the drop shadow to appear on screen. So we get the drawing order that v4 has with the good native canvas drop shadow looks of v5.

It doesn't solve 'double drop shadow with low alpha overlapping', but I think that'd require using a second canvas, which is out of scope for this PR

- [x] Documentation is changed or added
- [x] Lint process passed (`npm run lint`)
- [x] Tests passed (`npm run test`)

### CODE REVIEW

**Comment by themoonrat** [comment]
[v4](http://www.moonrat.co.uk/html5/pixi-drop-shadow/v4/#%7B%22style%22%3A%7B%22dropShadow%22%3Atrue%2C%22dropShadowAlpha%22%3A0.4%2C%22dropShadowAngle%22%3A1.5%2C%22dropShadowBlur%22%3A1%2C%22dropShadowColor%22%3A%22%2380ff00%22%2C%22dropShadowDistance%22%3A43%2C%22fill%22%3A%22red%22%2C%22fontSize%22%3A80%2C%22fontWeight%22%3A%22bold%22%2C%22letterSpacing%22%3A1%2C%22lineJoin%22%3A%22round%22%2C%22miterLimit%22%3A5%2C%22padding%22%3A10%2C%22stroke%22%3A%22yellow%22%2C%22strokeThickness%22%3A16%2C%22wordWrapWidth%22%3A0%7D%2C%22background%22%3A%22%23aaaaaa%22%7D)

[v5](http://www.moonrat.co.uk/html5/pixi-drop-shadow/v5/#%7B%22style%22%3A%7B%22dropShadow%22%3Atrue%2C%22dropShadowAlpha%22%3A0.4%2C%22dropShadowAngle%22%3A1.5%2C%22dropShadowBlur%22%3A1%2C%22dropShadowColor%22%3A%22%2380ff00%22%2C%22dropShadowDistance%22%3A43%2C%22fill%22%3A%22red%22%2C%22fontSize%22%3A80%2C%22fontWeight%22%3A%22bold%22%2C%22letterSpacing%22%3A1%2C%22lineJoin%22%3A%22round%22%2C%22miterLimit%22%3A5%2C%22padding%22%3A10%2C%22stroke%22%3A%22yellow%22%2C%22strokeThickness%22%3A16%2C%22wordWrapWidth%22%3A0%7D%2C%22background%22%3A%22%23aaaaaa%22%7D)

[v5-this-pr](http://www.moonrat.co.uk/html5/pixi-drop-shadow/v5-hybrid/#%7B%22style%22%3A%7B%22dropShadow%22%3Atrue%2C%22dropShadowAlpha%22%3A0.4%2C%22dropShadowAngle%22%3A1.5%2C%22dropShadowBlur%22%3A1%2C%22dropShadowColor%22%3A%22%2380ff00%22%2C%22dropShadowDistance%22%3A43%2C%22fill%22%3A%22red%22%2C%22fontSize%22%3A80%2C%22fontWeight%22%3A%22bold%22%2C%22letterSpacing%22%3A1%2C%22lineJoin%22%3A%22round%22%2C%22miterLimit%22%3A5%2C%22padding%22%3A10%2C%22stroke%22%3A%22yellow%22%2C%22strokeThickness%22%3A16%2C%22wordWrapWidth%22%3A0%7D%2C%22background%22%3A%22%23aaaaaa%22%7D)

**Comment by themoonrat** [comment]
Trying to fix with resolutions other than 1. *sigh*

File: packages/text/src/Text.js [hunk 1]
```diff
@@ -170,59 +170,79 @@ export default class Text extends Sprite
...
+            if (isShadowPass)
             {
-                linePositionX += maxLineWidth - lineWidths[i];
+                const dropShadowColor = style.dropShadowColor;
+                const rgb = hex2rgb(typeof dropShadowColor === 'number' ? dropShadowColor : string2hex(dropShadowColor));
+
+                context.shadowColor = `rgba(${rgb[0] * 255},${rgb[1] * 255},${rgb[2] * 255},${style.dropShadowAlpha})`;
+                context.shadowBlur = style.dropShadowBlur;
+                context.shadowOffsetX = Math.cos(style.dropShadowAngle) * style.dropShadowDistance;
+                context.shadowOffsetY = (Math.sin(style.dropShadowAngle) * style.dropShadowDistance) + dropShadowOffset;
```
Comment (TheWyo) [inline packages/text/src/Text.js]: Just came across this PR after spotting the rendering issue while playing around in the style editor. Multiplying dropShadowOffset by resolution here seems to sort it as far as I can tell (but not multiplying at dropShadowOffset's initialisation).

Comment (themoonrat) [reply packages/text/src/Text.js]: Literally just pushed that exact fix :D Great minds think alike!

Comment (TheWyo) [reply packages/text/src/Text.js]: Ha, great timing.

**Comment by TheWyo** [comment]
Hmm, bit more playing with this and changing the textBaseline (e.g. to top) now messes it up again by pulling all the text down, bringing the offscreen text for the shadow into view and nudging the main one off the bottom...

I've tried shifting the shadow pass off along the x instead of y which prevents that, but there's a little bit of duplicate text rendering on the left edge which I haven't figured out how to get rid of yet, thinking it might be padding related, just gotta find that last bit to tweak.

**Comment by themoonrat** [comment]
Do you think if I just make it double the height of the canvas, the text will be way way off screen and baseline won't bring it into view?

**Comment by TheWyo** [comment]
That's actually a much simpler solution yeah, seems to work as far as I can see.

**Comment by J77771977** [comment]
Thanks for commenting, We will revert back to you shortly.

**Comment by themoonrat** [comment]
Ok, made that change @TheWyo. You're testing on this is much appreciated; I'm worried there will be an edge case or a device or a browser that refuses to draw the shadow because the text itself is off-canvas. So far everywhere seems to work tho, so fingers crossed.

**Comment by TheWyo** [comment]
Yep, not seeing any shadow related issues anywhere on any of the different devices/browsers I've tested so far.

Though, I am getting some instances of text objects not rendering on iOS in the canvas renderer. I'm having difficulty reproducing it outside of the work project it's happening in, which annoyingly I can't share, and even then I'm not exactly sure what the trigger is that decides which objects it happens to. 

Going through the commits I've narrowed it down to that in [this one](https://github.com/pixijs/pixi.js/commit/0398354aa6bbd7c39d681f1990f0ded80c9449d2), the addition of the Math.ceil's in BaseTexture.setRealSize is what's causing it, though I have no idea why, or what's unique about the objects that are affected by it. I can only guess some edge-case about the render size of the text object?

**Comment by themoonrat** [comment]
I have another PR coming that might fix that. I went overzealous on the rounding; I'm assuming you have something not at a resolution of 1? Either the text resolution or the render resolution.

**Comment by TheWyo** [comment]
Yep, render resolution of 2.

**Comment by Bouh** [comment]
Spike on W cannot be fixed ? 
image

**Comment by TheWyo** [comment]
@Bouh That's related to your settings. Miter limit has to be a greater than zero value, zero itself is ignored by the canvas API. Change that (3 seems to look good), or use a different lineJoin type.

**Pull request merged**